File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        def helper(left: int, right: int) -> TreeNode:
            if left > right:
                return None
            
            nonlocal ppos
            root_idx = idx_map[preorder[ppos]]
            root = TreeNode(preorder[ppos])
            ppos += 1
            
            root.left = helper(left, root_idx - 1)
            root.right = helper(root_idx + 1, right)
            
            return root
        
        ppos = 0
        idx_map = {num: i for i, num in enumerate(inorder)}
        return helper(0, len(inorder) - 1)

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int preorderIdx;
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        this.preorderIdx = 0;
        
        Map<Integer, Integer> inorderIdxMap = new HashMap<>();
        for(int i = 0; i < inorder.length; i++)
            inorderIdxMap.put(inorder[i], i);
        
        return helper(preorder, inorderIdxMap, 0, inorder.length - 1);
    }
    
    private TreeNode helper(int[] preorder, Map<Integer, Integer> inorderIdxMap, int left, int right){
        if(left > right)
            return null;
        
        int rootIdx = inorderIdxMap.get(preorder[this.preorderIdx]);
        TreeNode root = new TreeNode(preorder[this.preorderIdx++]);
        
        root.left = helper(preorder, inorderIdxMap, left, rootIdx - 1);
        root.right = helper(preorder, inorderIdxMap, rootIdx + 1, right);
        
        return root;
    }
}